### ISSUE TITLE
ndk-build dependency tracking can be incorrect when pre-processing fails

### ISSUE THREAD

**Issue by b-spencer** [android-ndk/ndk#603]
#### Description

The Android NDK `ndk-build` build system doesn't delete the `.o` file if compilation fails during the pre-processing phase. When the `.o` file already existed before recompilation, this results in the old `.o` file remaining but no `.d` file existing, which means the existing out-of-date `.o` will not be properly rebuilt later (when only its `#include`-based dependencies have changed).  A workaround is to update the `ev-build-file` compilation command to touch the `.o` file first so that `.DELETE_ON_ERROR` always deletes the `.o` whenever the recipe for rebuilding an object fails.  Using `touch --no-create` is useful here because the file won't be created if it doesn't already exist.

It seems like the NDK makefiles should still have functional dependency tracking even when pre-processing fails.  I haven't been able to think of a downside for this workaround.  Perhaps the workaround is good enough that it can simply be made the built-in behaviour?

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: r13b, r15c, and r16 IIRC
* Build sytem: ndk-build
* Host OS: Debian 9 (Stretch)
* Compiler: N/A, but clang
* ABI: armeabi-v7a, x86_64
* STL: c++_shared
* NDK API level: N/A
* Device API level: N/A

**Comment by DanAlbert** [comment]
Tentatively r17, but if the fix is invasive then it'll be punted to r18.

**Comment by DanAlbert** [comment]
> When the .o file already existed before recompilation, this results in the old .o file remaining but no .d file existing

I see the opposite behavior. For a hello world ndk-build executable, I run:

```bash
$ ndk-build V=1
# See that the executable compiled and linked successfully.
$ ls obj/local/arm64-v8a/objs/foo/
foo.o foo.o.d
# Add an `#error foo` to foo.cpp
$ ndk-build V=1
# See compilation fail. No link.
$ ls obj/local/arm64-v8a/objs/foo/
foo.o.d
```

That's just a trivial `BUILD_EXECUTABLE` rule with a single foo.cpp with an empty main function. I'm using r19-canary (a fresh build of master), but I see the same with r13b and r15c.

Does that match what you see? I'm assuming that it's my guess at your project configuration that leads to the discrepancy, but it could be that this is a problem no longer exists (the documentation for `.DELETE_ON_ERROR` says that it will delete the file if the rule failed and the file was touched, maybe clang didn't used to touch the file if preprocessing failed and now it does), but given that I've tried master, r13, and r15, that seems unlikely.

**Comment by b-spencer** [comment]
Next week, I'll reproduce it locally again and dig a bit deeper to refresh my memory and report back.

FWIW, I do think it was an `#include` failure because of a missing file in the cases I original had.

**Comment by b-spencer** [comment]
My guess is that the `.d` file gets deleted when clang starts going through the `#include`s and then only written out (or only left written out) if that succeeds.  So, if you make it fail on `#error` perhaps you've already restored a (functional) `.d` by then and then things still work.

**Comment by b-spencer** [comment]
That seems to be the case:
```
$ ls
foo.cc  header.h
$ cat foo.cc
#include "header.h"
$ clang -MMD -c -o foo.o foo.cc
$ ls
foo.cc  foo.d  foo.o  header.h
$ rm header.h
$ clang -MMD -c -o foo.o foo.cc
foo.cc:1:10: fatal error: 'header.h' file not found
#include "header.h"
         ^~~~~~~~~~
1 error generated.
$ ls
foo.cc
```
(Although I see that the `.o` is removed too here, but I'm just using a random non-NDK version of clang here.)

Versus:
```
$ ls
foo.cc  header.h
$ clang -MMD -c -o foo.o foo.cc
$ ls
foo.cc  foo.d  foo.o  header.h
$ echo '#error bad' >>foo.cc
$ cat foo.cc
#include "header.h"
#error bad
$ clang -MMD -c -o foo.o foo.cc
foo.cc:2:2: error: bad
#error bad
 ^
1 error generated.
$ ls
foo.cc  foo.d  header.h
```

**Comment by b-spencer** [comment]
Here's a shell transcript that shows NDK r16b experiencing the original problem with an all-but-empty NDK project.  It seems that two layers of `#include` are necessary to make this fail.

```
$ tree
.
└── jni
    ├── Android.mk
    ├── Application.mk
    ├── first.h
    ├── program.c
    └── second.h

1 directory, 5 files
$ cat jni/Application.mk
APP_PLATFORM := android-24
APP_ABI := armeabi-v7a
$ cat jni/Android.mk
LOCAL_PATH := $(call my-dir)
include $(CLEAR_VARS)
LOCAL_MODULE := program
LOCAL_SRC_FILES := program.c
include $(BUILD_STATIC_LIBRARY)
$ cat jni/program.c
#include "first.h"
$ cat jni/first.h
#include "second.h"
$ cat jni/second.h
$ ndk-build
[armeabi-v7a] Compile thumb  : program <= program.c
[armeabi-v7a] StaticLibrary  : libprogram.a
$ tree
.
├── jni
│   ├── Android.mk
│   ├── Application.mk
│   ├── first.h
│   ├── program.c
│   └── second.h
└── obj
    └── local
        └── armeabi-v7a
            ├── libprogram.a
            └── objs
                └── program
                    ├── program.o
                    └── program.o.d

6 directories, 8 files
$ ndk-build
$ rm jni/second.h
$ ndk-build
[armeabi-v7a] Compile thumb  : program <= program.c
In file included from jni/program.c:1:
jni/first.h:1:10: fatal error: 'second.h' file not found
#include "second.h"
         ^~~~~~~~~~
1 error generated.
make: *** [obj/local/armeabi-v7a/objs/program/program.o] Error 1
$ tree
.
├── jni
│   ├── Android.mk
│   ├── Application.mk
│   ├── first.h
│   └── program.c
└── obj
    └── local
        └── armeabi-v7a
            ├── libprogram.a
            └── objs
                └── program
                    └── program.o

6 directories, 6 files
$ ndk-build ########## This should fail but it doesn't!
$ tree
.
├── jni
│   ├── Android.mk
│   ├── Application.mk
│   ├── first.h
│   └── program.c
└── obj
    └── local
        └── armeabi-v7a
            ├── libprogram.a
            └── objs
                └── program
                    └── program.o

6 directories, 6 files
$ rm -rf obj/  ###### Forcing recompilation shows that it still fails.
$ ndk-build
[armeabi-v7a] Compile thumb  : program <= program.c
In file included from jni/program.c:1:
jni/first.h:1:10: fatal error: 'second.h' file not found
#include "second.h"
         ^~~~~~~~~~
1 error generated.
make: *** [obj/local/armeabi-v7a/objs/program/program.o] Error 1
$
```

Making the change I outline in the issue description fixes this behaviour so that the subsequent `ndk-build` invocations also fail (because `program.o` is always deleted by make).

**Comment by DanAlbert** [comment]
Great, thanks for the detailed repro case :+1:

I should be able to find some time to look at this on Friday.

**Comment by DanAlbert** [comment]
tl;dr: r17+ work by accident (assuming you haven't manually added `-fno-integrated-as`, real fix will be in r19.

Was able to reproduce this with r16b, but not with r19 or even r17b. The difference between the two is that after the expected compilation failure we get after removing second.h, r16 still has a program.o:

```bash
# With r16b
$ ndk-build
[armeabi-v7a] Compile thumb  : program <= program.c
In file included from jni/program.c:1:
jni/first.h:1:10: fatal error: 'second.h' file not found
#include "second.h"    
         ^~~~~~~~~~         
1 error generated.
make: *** [obj/local/armeabi-v7a/objs/program/program.o] Error 1
$ tree
.
├── jni
│   ├── Android.mk
│   ├── Application.mk
│   ├── first.h
│   └── program.c
└── obj
    └── local
        └── armeabi-v7a
            ├── libprogram.a
            └── objs
                └── program
                    └── program.o
# With r17b
$ ndk-build
[armeabi-v7a] Compile thumb  : program <= program.c
In file included from jni/program.c:1:
jni/first.h:1:10: fatal error: 'second.h' file not found
#include "second.h"
         ^~~~~~~~~~
1 error generated.
make: *** [obj/local/armeabi-v7a/objs/program/program.o] Error 1
$ tree
.
├── jni
│   ├── Android.mk
│   ├── Application.mk
│   ├── first.h
│   └── program.c
└── obj
    └── local
        └── armeabi-v7a
            ├── libprogram.a
            └── objs
                └── program
```

Looks like the fix was, surprisingly, https://android-review.googlesource.com/c/platform/ndk/+/622415

```bash
$ touch foo.o
$ echo '#error foo' > foo.c
$ clang -c foo.c -o foo.o -fno-integrated-as
foo.c:1:2: error: foo
#error foo
 ^
1 error generated.
$ ls
foo.c foo.o
$ clang -c foo.c -o foo.o -fintegrated-as
foo.c:1:2: error: foo
#error foo
 ^
1 error generated.
$ ls
foo.c
```

It looks like this is also gcc's behavior.

If I add

```makefile
APP_CFLAGS := -fno-integrated-as
```

to your test case's Application.mk the issue returns, even with r19. ndk-build does use `.DELETE_ON_ERROR` which is supposed to prevent this kind of thing, but apparently that only takes effect if the rule fails *and* the output changes at all. I think the problem is actually that the deps file is considered an implicit output of the rule and it *does* change, so it's deleted. When run again, make doesn't find the dependency file so it believes the object file is up to date and doesn't bother rerunning.

https://android-review.googlesource.com/c/platform/ndk/+/793957 fixes this regardless of which assembler is used.

Thanks again for the detailed repro case. That made this much easier to diagnose and fix.

**Comment by DanAlbert** [comment]
Since I have a fix for this, will pull in to beta 1 after all.

**Comment by rprichard** [comment]
It looks like GCC leaves the object file and depfile untouched if the compile command fails.

Clang, on the other hand, deletes the depfile on a missing include error, and otherwise updates it. Its behavior for the object file depends on `-f[no-]integrated-as`:
 - `-fintegrated-as`: delete the object file on any compile error
 - `-fno-integrated-as`: leave the old file untouched on an error

The GCC behavior seems fine to me, because a future `make` invocation should retry the failed compile step. The Clang `-fno-integrated-as` behavior, on the other hand, seems broken to me. It's deleting the depfile but not the object file, so future builds reuse the stale object file.

The CL seems fine to me as a fix though.

**Comment by DanAlbert** [comment]
I spoke with ccross briefly, and from the sound of things tool owners and build system owners (as in ninja and make, not ndk-build) disagree about whose job it is to cleanup outputs when a command fails, so while I agree that clang's behavior ought to be better here, they might not share that opinion. We probably want this patch regardless since fixing clang only fixes clang. I don't know if yasm, renderscript, etc might have similar issues.

**Comment by b-spencer** [comment]
Interesting.  FWIW, I've never fully understood why make's `.DELETE_ON_ERROR` requires the target to change before it will act.  I can't think of a time when an out-of-date target (and thus a target being rebuilt) is so valuable that it needs to be kept when an error occurs during its recipe.  (There don't seem to be any hints in the project history and I can't find a mailing list or any other discussions for make from 1994, when the feature was added.)

**Comment by DanAlbert** [comment]
Yeah, it seems like a strange decision to me too.

Fix is in master right now and presubmit is working on the cherry-pick to r19, so this will be in r19 beta 1. Will leave this open until it's actually submitted just so I don't forget to make sure it actually lands.
